components/calendar/src/any_calendar.rs

```diff
@@ -0,0 +1,324 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+//! Module for working with multiple calendars at once
+
+use crate::buddhist::Buddhist;
+use crate::coptic::Coptic;
+use crate::gregorian::Gregorian;
+use crate::indian::Indian;
+use crate::iso::Iso;
+use crate::japanese::Japanese;
+
+use crate::{types, Calendar, Date, DateDuration, DateDurationUnit};
+
+/// This is a calendar that encompasses all formattable calendars supported by this crate
+///
+/// This allows for the construction of [`Date`] objects that have their calendar known at runtime.
+pub enum AnyCalendar {
+    Gregorian(Gregorian),
+    Buddhist(Buddhist),
+    Japanese(Japanese),
+    Indian(Indian),
+    Coptic(Coptic),
+    Iso(Iso),
+}
+
+/// The inner date type for [`AnyCalendar`]
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub enum AnyDateInner {
+    Gregorian(<Gregorian as Calendar>::DateInner),
+    Buddhist(<Buddhist as Calendar>::DateInner),
+    Japanese(<Japanese as Calendar>::DateInner),
+    Indian(<Indian as Calendar>::DateInner),
+    Coptic(<Coptic as Calendar>::DateInner),
+    Iso(<Iso as Calendar>::DateInner),
+}
+
+macro_rules! match_cal_and_date {
+    (match ($cal:ident, $date:ident): ($cal_matched:ident, $date_matched:ident) => $e:expr) => {
+        match ($cal, $date) {
+            (&Self::Gregorian(ref $cal_matched), &AnyDateInner::Gregorian(ref $date_matched)) => $e,
+            (&Self::Buddhist(ref $cal_matched), &AnyDateInner::Buddhist(ref $date_matched)) => $e,
+            (&Self::Japanese(ref $cal_matched), &AnyDateInner::Japanese(ref $date_matched)) => $e,
+            (&Self::Indian(ref $cal_matched), &AnyDateInner::Indian(ref $date_matched)) => $e,
+            (&Self::Coptic(ref $cal_matched), &AnyDateInner::Coptic(ref $date_matched)) => $e,
+            (&Self::Iso(ref $cal_matched), &AnyDateInner::Iso(ref $date_matched)) => $e,
+            _ => panic!(
+                "Found AnyCalendar with mixed calendar type {} and date type {}!",
```

Comment:
Discussion: I guess `panic` is the right thing to do here? Did you consider alternatives for handling mismatched calendar arguments?

Reply:
Yes: I considered Results, but that would impact the API for all other calendars for errors that only really make sense for AnyCalendar. The only way to trigger this is to manually construct dates from `from_raw()` and users aren't supposed to use that API unless they're implementing their own calendars